Topics/Development_Process/Duckdns.md

```diff
@@ -0,0 +1,53 @@
+# Overcoming a Dynamic IP Address with DuckDNS
+
+## Motivation
+
+So you have chosen to deploy your application on AWS or perhaps you have turned a spare computer of yours into your home server. You are finally deployed and users can finally access your site! I mean, the address for your site may be something weird like `http://123.0.0.321` but at least _someone_ can access your site right?
+
+Well, if you ever have to restart your server to make updates or save some compute power when nobody is using your server, you may find that the next time you boot the server up, the IP address of your deployment has changed! Now you have to tell everyone your application has a new URL. This is because most hosting options (especially the free/low cost options) give your production server what is called a _Dynamic IP address_. In other words, services like AWS or your Internet Service Provider assign a new public (or external) IP address to your server whenever it boots up and joins the network. This is often done to save resources as an IP not used by one person can be used by another.
+
+While you could contact your hosting service or ISP to make it so that the external IP address of your server, which everyone connects to, stays fixed, this can either be quite a hassle or cost you money (something most students don't have to spare!). To get around this problem, one can turn to a _Dynamic DNS_ provider. A popular choice of such a provider, among amateurs and hobbyists, is a service called DuckDNS.
+
+## What is DuckDNS?
+
+DuckDNS is a Dynamic Domain Name Service provider. In other words, they are a service that maps domain names to dynamic IP addresses. When using DuckDNS, your application gets a domain name. As it is a free service however, this name will be a subdomain of DuckDNS (e.g. `http://my-cool-application.duckdns.org`). More importantly however, this name will not change. DuckDNS then provides a script you run on your server that tells DuckDNS what the external IP address of your server is. This script is configured by default to run every 5 minutes.
+
+To summarize what DuckDNS does, whenever a user tries to navigate to your application, as part of resolving the domain name of your application, the user asks DuckDNS what the IP address is of your application. Since you constantly update DuckDNS with your application's IP address, the user is up to date with your server's address even if it changes.
+
+![Image to illustrate how Dynamic DNS works](https://learningtechnix.files.wordpress.com/2020/04/ddns_image.png)
+
+## How do I use DuckDNS?
+
+Using DuckDNS is very straightforward. You can sign up by logging into your twitter, github, or google account. You are then taken to a homepage that shows your account token and the domain name mappings you created. Simply create a new domain name for your application. You do not need to set your application's IP address on this page (the script that will be run does that for us).
```

Comment:
Possibly mention the importance of keeping your account token secure and possibly hidden. If I understood correctly, the account token acts as a unique identifier for your DuckDNS account. Thus if it falls into the wrong hands, unauthorized individuals might gain access to your account and make changes without your permission.